src/Composer/Installer.php

```diff
@@ -0,0 +1,209 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\ManagerPlugin\Composer;
+
+use Composer\Package\Locker;
+use Composer\Package\RootPackageInterface;
+use Contao\ManagerPlugin\Dependency\DependencyResolverTrait;
+use Contao\ManagerPlugin\Dependency\DependentPluginInterface;
+use Contao\ManagerPlugin\Dependency\UnresolvableDependenciesException;
+
+class Installer
+{
+    use DependencyResolverTrait;
+
+    private static $generatedClassTemplate = <<<'PHP'
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\ManagerPlugin;
+
+/**
+ * This class is generated by contao/manager-bundle, specifically by
+ * @see \Contao\ManagerBundle\Composer\Installer
+ *
+ * This file is overwritten at every run of `composer install` or `composer update`.
+ */
+%s
+{
+    const BUNDLE_PLUGINS = 'Contao\ManagerPlugin\Bundle\BundlePluginInterface';
+    const CONFIG_PLUGINS = 'Contao\ManagerPlugin\Config\ConfigPluginInterface';
+    const EXTENSION_PLUGINS = 'Contao\ManagerPlugin\Config\ExtensionPluginInterface';
+    const ROUTING_PLUGINS = 'Contao\ManagerPlugin\Routing\RoutingPluginInterface';
+
+    /**
+     * @var array
+     */
+    private $plugins;
+
+    /**
+     * Constructor.
+     */
+    public function __construct()
+    {
+        $this->plugins = %s;
+    }
+
+    /**
+     * Returns all active plugin instances.
+     *
+     * @return array
+     */
+    public function getInstances()
+    {
+        return $this->plugins;
+    }
+
+    /**
+     * Returns the active plugin instances of a given type (see class constants).
+     *
+     * @param string $type
+     * @param bool   $reverseOrder
+     *
+     * @return array
+     */
+    public function getInstancesOf($type, $reverseOrder = false)
+    {
+        $plugins = array_filter(
+            $this->getInstances(),
+            function ($plugin) use ($type) {
+                return is_a($plugin, $type);
+            }
+        );
+
+        $plugins = $reverseOrder ? array_reverse($plugins, true) : $plugins;
+
+        return $plugins;
+    }
+}
+
+PHP;
+
+    /**
+     * Sets the Contao Manager plugins.
+     *
+     * @param Locker               $locker
+     * @param RootPackageInterface $rootPackage
+     *
+     * @throws \RuntimeException
+     */
+    public function dumpPlugins(Locker $locker, RootPackageInterface $rootPackage)
+    {
+        $plugins = [];
+
+        $disabled = $this->getDisabledPackages($rootPackage);
+        $lockData = $locker->getLockData();
+        $lockData['packages-dev'] = isset($lockData['packages-dev']) ? $lockData['packages-dev'] : [];
```

Comment:
What's the point of this line?

Reply:
Include the dev packages? This is copied from `package-versions`

Reply:
This is a strange way of putting it. I guess it translates to:

```php
if (!isset($lockData['packages-dev'])) {
    $lockData['packages-dev'] = [];
}
```

Reply:
yeah… do I need to change it?

Reply:
Depends on whether or not you want people to think "this guy writes weird code". 😄